### ISSUE TITLE
torch._weight_norm with specified dim returns wrong output

### ISSUE THREAD

**Issue by xuzijian629** [pytorch/pytorch#81195]
### 🐛 Describe the bug

# Problem
`torch._weight_norm` outputs wrong result from torch 1.12

Though `torch._weight_norm` is an undocumented API, I report this issue since the API is made public.

#### torch 1.11 and before
```python3
$ python3
Python 3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> a = torch.Tensor([[1,2],[3,4]])
>>> b = torch.Tensor([[5],[6]])
>>> torch._weight_norm(a,b,dim=1)
tensor([[1.5811, 2.2361],
        [5.6921, 5.3666]])
```

#### torch 1.12
```python3
$ python3
Python 3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> a = torch.Tensor([[1,2],[3,4]])
>>> b = torch.Tensor([[5],[6]])
>>> torch._weight_norm(a,b,dim=1)
tensor([[1.5811, 2.6833],
        [4.7434, 5.3666]])
```

#### manual computation
```python3
>>> def weight_norm_dim1(a, b):
...     norm = a.norm(2, [0], True)
...     return a * b / norm
... 
>>> weight_norm_dim1(a, b)
tensor([[1.5811, 2.2361],
        [5.6921, 5.3666]])
```

### Versions

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0
Clang version: Could not collect
CMake version: version 3.22.2
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-100-generic-x86_64-with-glibc2.2.5
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 470.103.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] torch==1.12.0
[pip3] torchvision==0.13.0
[conda] Could not collect
```

**Comment by jbschlosser** [comment]
Due to #73845?

**Comment by albanD** [comment]
@frank-wei can you take a look at this?

**Comment by albanD** [comment]
Also this should have been caught by tests.

**Comment by frank-wei** [comment]
cc @mingfeima @jgong5

**Comment by mingfeima** [comment]
It should be torch 1.11 and before got the wrong result on CPU. The difference here is on the CPU side add a fused path when dim is the first dimension or the last dimension, before that CPU would go the non-fused path.

CUDA result (same as torch 1.12 on CPU):

```
version: torch 1.11.0+cu115
Python 3.8.13 (default, Mar 28 2022, 11:38:47)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> a = torch.Tensor([[1,2],[3,4]]).cuda()
>>> b = torch.Tensor([[5],[6]]).cuda()
>>> torch._weight_norm(a,b,dim=1)
tensor([[1.5811, 2.6833],
        [4.7434, 5.3666]], device='cuda:0')
```


`weight_norm` shape for `v` and `g` (`a` and `b` in the test above) is defined from this page https://pytorch.org/docs/stable/generated/torch.nn.utils.weight_norm.html

when dim is the last dimension, `v` has shape of [M, N] and `g` has shape of [1, N], for example:
```
>>> import torch
>>> from torch import nn
>>> m = torch.nn.utils.weight_norm(nn.Linear(20, 40), name='weight', dim=1)
>>> m.weight_v.size()
torch.Size([40, 20])
>>> m.weight_g.size()
torch.Size([1, 20])
```

In the reported example if we change `b = torch.Tensor([[5, 6]])`, it would also get the result of
```
tensor([[1.5811, 2.2361],
        [5.6921, 5.3666]])
```

`norm` has shape of [1, N], if `b` is given with shape of [N, 1], `g / norm` would have the shape of [N, N] instead of [1, N], so the final result is not correct.

**Comment by mingfeima** [comment]
@frank-wei could you please also check CUDA result of the reported example on your side?

**Comment by albanD** [comment]
Good catch @mingfeima !
There are indeed invalid inputs for this function, so it is ok to return arbitrary results. The public `torch.nn.utils.weight_norm()` function would never generate such input.

This private function could do more error checking but I don't think we want to require it?

**Comment by mingfeima** [comment]
> Good catch @mingfeima ! There are indeed invalid inputs for this function, so it is ok to return arbitrary results. The public `torch.nn.utils.weight_norm()` function would never generate such input.
> 
> This private function could do more error checking but I don't think we want to require it?

I think the same, as long as `weight_norm` is bonded on other modules like `nn.Linear` or `nn.LSTM`, it would always have inputs in the correct shape.

**Comment by frank-wei** [comment]
Should we close this issue @albanD ?

**Comment by xuzijian629** [comment]
Thanks for the investigation!
Yes, I think we can close this issue.

**Comment by albanD** [comment]
Given that it is easy to check, I think it would be worth it to add a check in the c++ function make sure this doesn't happen anymore.

**Comment by frank-wei** [comment]
Agree. A shape check should be ok I think.
